### PR TITLE
Make apply_light_arc() use castLight()

### LINKED ISSUE #12182: Reimplement dynamic lighting using castLight().
The code that applies dynamic lighting is still using the old manual raytracing code from olden times, which especially for bright lights is horrifically slow.
One should be able to reimplement apply_light_arc() as a method that slices up the arc into octants and partial octants and feeds them to castLight, which is guaranteed to only visit each square being lit once and handles a lot of tricky trig and light absorption stuff.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/12123389-reimplement-dynamic-lighting-using-castlight?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by WelpThatWorked**
#### Summary
Performance "Improve lighting performance"

#### Purpose of change

Fixes #12182, #2415

#### Describe the solution

Rewrite apply_light_arc() to use castLight(). Add cot() function for units::angle.

#### Testing

Spawned Survivor RV, drove in circles with headlights on (90 degree beam), using some simple timers to measure performance difference. On my laptop, previous code took 0.45-1.4 ms per call, new code takes 0.3 to 0.08 ms per call. 

#### Additional context

I have not written much code with high performance in mind before, so there probably is something I could do better I'm missing.

### CODE REVIEW

File: src/lightmap.cpp [hunk 1]
```diff
@@ -46,6 +46,7 @@
 #include "vpart_range.h"
 #include "weather.h"
 #include "weather_type.h"
+#include <units_utility.h>
```
Comment (BrettDong) [inline src/lightmap.cpp:49]: Please use quotes (`#include "units_utility.h"`) or it doesn't compile, and insert it after `#include "units.h"`, so headers are listed in lexicographical order.

File: src/lightmap.cpp [hunk 2]
```diff
@@ -1249,46 +1250,124 @@ void map::apply_light_arc( const tripoint &p, const units::angle &angle, float l
...
+            }
+        }
+        int numoct = to_degrees( cangle - oangle ) / 45;
+        int firstoct = to_degrees( oangle ) / 45;
+        oangle += numoct * 45_degrees;
+        wangle = cangle - oangle;
+
+        for( int i = firstoct; i < numoct + firstoct; i++ ) {
+            //if arc crosses 0 degrees, i.e. sectors 7-0-1, offset back to sector 0 after 7
+            switch( i > 7 ? i - 8 : ( i < 0 ? i + 8 : i ) ) {
```
Comment (BrettDong) [inline src/lightmap.cpp:1311]: This expression can be simplified
```suggestion
            switch( ( i + 8 ) % 8 ) {
```

File: src/lightmap.cpp [hunk 3]
```diff
@@ -1249,46 +1250,124 @@ void map::apply_light_arc( const tripoint &p, const units::angle &angle, float l
...
+            for( units::angle ao = 0_degrees; ao <= preanglew; ao += wstep ) {
+                if( trigdist ) {
+                    double fdist = ( ao * M_PI_2 ) / preanglew;
+                    end.x = static_cast<int>(
+                                p.x + ( static_cast<double>( range ) - fdist * 2.0 ) * cos( preangle + ao ) );
+                    end.y = static_cast<int>(
+                                p.y + ( static_cast<double>( range ) - fdist * 2.0 ) * sin( preangle + ao ) );
+                    apply_light_ray( lit, p, end, luminance );
+                } else {
+                    calc_ray_end( nangle + ao, range, p, end );
```
Comment (BrettDong) [inline src/lightmap.cpp:1299]: Did you mean `preangle + ao` here?

File: src/lightmap.cpp [hunk 4]
```diff
@@ -1249,46 +1250,124 @@ void map::apply_light_arc( const tripoint &p, const units::angle &angle, float l
...
+                    apply_light_ray( lit, p, end, luminance );
+                }
+            }
+        }
+        int numoct = to_degrees( cangle - oangle ) / 45;
+        int firstoct = to_degrees( oangle ) / 45;
```
Comment (BrettDong) [inline src/lightmap.cpp:1305]: To get rid of type casting warnings
```suggestion
        int numoct = std::floor( to_degrees( cangle - oangle ) / 45 );
        int firstoct = static_cast<int>( std::lround( to_degrees( oangle ) / 45 ) );
```

File: src/lightmap.cpp [hunk 5]
```diff
@@ -1249,46 +1250,124 @@ void map::apply_light_arc( const tripoint &p, const units::angle &angle, float l
...
+                                  lm, transparency_cache, p2, 0, luminance );
+                    break;
+                case 3:
+                    castLight < 0, 1, -1, 0, float, four_quadrants, light_calc, light_check,
+                              update_light_quadrants, accumulate_transparency > (
+                                  lm, transparency_cache, p2, 0, luminance );
+                    break;
+                case 4:
+                    castLight<0, 1, 1, 0, float, four_quadrants, light_calc, light_check,
+                              update_light_quadrants, accumulate_transparency>(
```
Comment (BrettDong) [inline src/lightmap.cpp:1334]:
```suggestion
                    castLight < 0, 1, 1, 0, float, four_quadrants, light_calc, light_check,
                              update_light_quadrants, accumulate_transparency > (
```

File: src/lightmap.cpp [hunk 6]
```diff
@@ -1249,46 +1250,124 @@ void map::apply_light_arc( const tripoint &p, const units::angle &angle, float l
...
+                                  lm, transparency_cache, p2, 0, luminance );
+                    break;
+                case 5:
+                    castLight<1, 0, 0, 1, float, four_quadrants, light_calc, light_check,
+                              update_light_quadrants, accumulate_transparency>(
```
Comment (BrettDong) [inline src/lightmap.cpp:1339]:
```suggestion
                    castLight < 1, 0, 0, 1, float, four_quadrants, light_calc, light_check,
                              update_light_quadrants, accumulate_transparency > (
```

File: src/lightmap.cpp [hunk 7]
```diff
@@ -1249,46 +1250,124 @@ void map::apply_light_arc( const tripoint &p, const units::angle &angle, float l
...
 
     tripoint end;
     int range = LIGHT_RANGE( luminance );
     calc_ray_end( nangle, range, p, end );
-    apply_light_ray( lit, p, end, luminance );
 
     tripoint test;
     calc_ray_end( wangle + nangle, range, p, test );
 
     const float wdist = hypot( end.x - test.x, end.y - test.y );
```
Comment (BrettDong) [inline src/lightmap.cpp:1273]: BTW although not involved in this pull request, `wdist` should be declared `double` to save single/double conversions
```suggestion
    const double wdist = hypot( end.x - test.x, end.y - test.y );
```
currently the return value from `hypot` is converted to single precision and then compared to `0.5`:
```assembly
    0x103bc0470 <+380>:  bl     0x103bc97a4               ; hypot<int, int> at math.h:1267
    0x103bc0474 <+384>:  fcvt   s0, d0
    0x103bc0478 <+388>:  fmov   s1, #0.50000000
    0x103bc047c <+392>:  fcmp   s0, s1
```

**Comment by anothersimulacrum** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/issues/2415 is also relevant?

**Comment by jbytheway** [comment]
In principle I think we should be able to do even better than this.  The remaining portions of the arc that aren't whole eighths should also be doable using the shadowcasting algorithm.  But maybe the current interface is not sufficient to permit that.  In any case, this PR looks like a step in the right direction; no need to avoid merging it just because something better might be possible.

**Comment by WelpThatWorked** [comment]
@jbytheway Your comment inspired me to take a second look and figure out how to use castLight on the partial sections, as previously I just assumed it couldn't be done. Thanks.

**Comment by kevingranade** [comment]
Seems about 90% of the way there, but there's a bit of skew.
Screenshot from 2021-07-10 18-13-26

**Pull request merged**